src/drivers/gitea.go

```diff
@@ -0,0 +1,195 @@
+package drivers
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/git-town/git-town/src/drivers/helpers"
+	"golang.org/x/oauth2"
+
+	"code.gitea.io/sdk/gitea"
+)
+
+// GiteaConfig defines the data that the giteaCodeHostingDriver needs from the Git configuration.
+type GiteaConfig interface {
+	GetCodeHostingDriverName() string
+	GetRemoteOriginURL() string
+	GetGiteaToken() string
+	GetCodeHostingOriginHostname() string
+}
+
+type giteaCodeHostingDriver struct {
+	originURL  string
+	hostname   string
+	apiToken   string
+	client     *gitea.Client
+	owner      string
+	repository string
+}
+
+// GiteaConfig provides a Gitea driver instance if the given repo configuration is for a Gitea repo,
+// otherwise nil.
+func LoadGitea(config GiteaConfig) CodeHostingDriver {
+	driverType := config.GetCodeHostingDriverName()
+	originURL := config.GetRemoteOriginURL()
+	hostname := helpers.GetURLHostname(originURL)
+	configuredHostName := config.GetCodeHostingOriginHostname()
+	if configuredHostName != "" {
+		hostname = configuredHostName
+	}
+	if driverType != "gitea" && hostname != "gitea.com" {
+		return nil
+	}
+	repositoryParts := strings.SplitN(helpers.GetURLRepositoryName(originURL), "/", 2)
+	if len(repositoryParts) != 2 {
+		return nil
+	}
+	owner := repositoryParts[0]
+	repository := repositoryParts[1]
+	return &giteaCodeHostingDriver{
+		originURL:  originURL,
+		hostname:   hostname,
+		apiToken:   config.GetGiteaToken(),
+		owner:      owner,
+		repository: repository,
+	}
+}
+
+func (d *giteaCodeHostingDriver) CanMergePullRequest(branch, parentBranch string) (canMerge bool, defaultCommitMessage string, pullRequestNumber int64, err error) {
+	if d.apiToken == "" {
+		return false, "", 0, nil
+	}
+	d.connect()
+	openPullRequests, err := d.client.ListRepoPullRequests(d.owner, d.repository, gitea.ListPullRequestsOptions{
+		ListOptions: gitea.ListOptions{
+			PageSize: 50,
+		},
+		State: gitea.StateOpen,
+	})
+	if err != nil {
+		return false, "", 0, err
+	}
+	fmt.Printf("1111 %#v\n", openPullRequests[0])
+	baseName := parentBranch
+	headName := d.owner + "/" + branch
+	pullRequests := filterPullRequests(openPullRequests, baseName, headName)
+	if len(pullRequests) != 1 {
+		return false, "", 0, nil
+	}
+	pullRequest := pullRequests[0]
+	if !pullRequest.Mergable {
+		return false, "", 0, nil
+	}
```

Comment:
@kevgo I'd like to tell the user that the server reports it indeed has found a commit, but this commit is not mergeable, "please open the following url to check: ...". However it looks like this would be silently ignored in the current implementation.

I'd like to suggest a `LoadPullRequest()` which should return a user exposed error, and otherwise load all state (and keep it on the driver) for doing the pull request. 

https://github.com/git-town/git-town/blob/8cd928ed8952bbfd897646006a98f851d72c1eda/src/cmd/ship.go#L68-L69

Could become:
```golang
		driver := drivers.Load(repo.Configuration)
		config, err := gitShipConfig(args, driver)
		ok, err := driver.InitMergePullRequest(config) 
		if err != nil {
                         // tell the user why not and eventually go on
                }
                config.canShipWithDriver = ok
		if err != nil {
```

Reply:
Helpful error messages to the user are always welcome! Let's build that as a separate feature. Maybe as a `canShip(config)` method that returns a string. If the string is empty, it means no problem. If the string contains something, it is considered an error message, printed to the user, and the program exits.